File: src/clase-05/typescript/decorator.ts
// Componente
export interface INotifier {
  send(message: string): string;
}

// Componente concreto
export class Notifier implements INotifier {
  send(message: string): string {
    return `Enviando: ${message}`;
  }
}

// Decorador base
export class NotifierDecorator implements INotifier {
  constructor(protected readonly notifier: INotifier) {}

  send(message: string): string {
    return this.notifier.send(message);
  }
}

// Decoradores concretos
export class EmailDecorator extends NotifierDecorator {
  send(message: string): string {
    return `${super.send(message)} | canal=Email`;
  }
}

export class SMSDecorator extends NotifierDecorator {
  send(message: string): string {
    return `${super.send(message)} | canal=SMS`;
  }
}

export class SlackDecorator extends NotifierDecorator {
  send(message: string): string {
    return `${super.send(message)} | canal=Slack`;
  }
}

if (require.main === module) {
  const base = new Notifier();
  const multi = new SlackDecorator(new SMSDecorator(new EmailDecorator(base)));
  // eslint-disable-next-line no-console
  console.log(multi.send("Hola equipo"));
}

File: src/clase-05/python/decorator.py
from __future__ import annotations

# Componente base
class Notifier:
    def send(self, message: str) -> str:
        return f"Enviando: {message}"

# Decorador base
class NotifierDecorator(Notifier):
    def __init__(self, notifier: Notifier):
        self._notifier = notifier

    def send(self, message: str) -> str:
        return self._notifier.send(message)

# Decoradores concretos
class EmailDecorator(NotifierDecorator):
    def send(self, message: str) -> str:
        base = super().send(message)
        return f"{base} | canal=Email"

class SMSDecorator(NotifierDecorator):
    def send(self, message: str) -> str:
        base = super().send(message)
        return f"{base} | canal=SMS"

class SlackDecorator(NotifierDecorator):
    def send(self, message: str) -> str:
        base = super().send(message)
        return f"{base} | canal=Slack"

def main():
    base = Notifier()
    multi = SlackDecorator(SMSDecorator(EmailDecorator(base)))
    print(multi.send("Hola equipo"))

if __name__ == "__main__":
    main()
